File: Java/P2/irrgaten/src/irrgarten/Labyrinth.java

package irrgarten;
import java.util.ArrayList;

/**
 * @brief   Funciones del Laberinto
 *          EL (0,0) ES LA POSICIÓN SUPERIOR IZQUIERDA --> LAS FILAS SON NEGATIVAS
 *                                                         Y LAS COLUMNAS POSITIVAS                                                                          
 * @author  Isabel Morro Tabares
 */
 

public class Labyrinth {
    
    // Declaración de atributos
    
    private static final char BLOCK_CHAR = 'X';
    private static final char EMPTY_CHAR = '-';
    private static final char MONSTER_CHAR = 'M';
    private static final char COMBAT_CHAR = 'C';
    private static final char EXIT_CHAR = 'E';
    private static final int ROW = 0;
    private static final int COL = 1;
    
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;

   
    private Monster[][] monsters;
    private char[][] labyrinth;
    private Player[][] players;
 
    
    // Constructor
    
    public Labyrinth (int nRows, int nCols, int exitRow, int exitCol){
        
        this.nCols = nCols;
        this.nRows = nRows;
        this.exitCol = exitCol;
        this.exitRow = exitRow;
        
        monsters = new Monster [nRows][nCols];
        labyrinth = new char [nRows][nCols];
        players = new Player[nRows][nCols];
        
        
        
        // Casllas vacías y muros
        for (int i = 0; i < nRows ; i++){
            for (int j = 0; j < nCols; j++){
                if ((i == 0 || i == nRows-1 || j == 0 || j == nCols-1)){
                    labyrinth[i][j] = BLOCK_CHAR; 
                }
                else {
                    labyrinth[i][j] = EMPTY_CHAR; 
                }
            }
        }
        
        // Casilla de salida
        labyrinth [this.exitRow + nRows - 1][this.exitCol] = EXIT_CHAR;
        
    }
    
    // Métodos privados
    
    /**
     * @brief Comprueba que la posición está dentro del laberinto
     * @param row Fila
     * @param col Columna
     * @return true si está dentro, false en caso contrario
     */
    private boolean posOK (int row, int col){
        return -row < nRows && col < nCols && row <= 0 && col >= 0;
    }
    
    /**
     * @brief Comprueba si la posición es vacía
     * @param row Fila
     * @param col Columna
     * @return True si vacía, false en caso contrario
     */
    private boolean emptyPos (int row, int col){
        return labyrinth[nRows+row-1][col] == EMPTY_CHAR;
    }
    
    /**
     * @brief Comprueba si la posición contiene un monstruo
     * @param row Fila
     * @param col Columna
     * @return True si lo tiene, false en caso contrario
     */
    private boolean monsterPos (int row, int col){
        return labyrinth[nRows+row-1][col] == MONSTER_CHAR;
    }
    
    /**
     * @brief Comprueba si la posición es la de salida
     * @param row Fila
     * @param col Columna
     * @return True si lo es, false en caso contrario
     */
    private boolean exitPos (int row, int col){
        return labyrinth[nRows+row-1][col] == EXIT_CHAR;
    }
    
    /**
     * @brief Comprueba si en la posición hay un combate
     * @param row Fila
     * @param col Columna
     * @return True si lo hay, false en caso contrario
     */
    private boolean combatPos (int row, int col){
        return labyrinth[nRows+row-1][col] == COMBAT_CHAR;
    }
    
    /**
     * @brief Indica si la posicion es correcta y, o es vacía, o hay un monstruo,
     *        o es la casilla de salida
     * @param row Fila
     * @param col Columna
     * @return True si lo es, False en caso contrario
     */
    private boolean canStepOn (int row, int col){
        return posOK(row,col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col));
    }
    
    /**
     * @brief Actualiza la casilla
     * @param row Fila
     * @param col Columna
     */
    private void updateOldPos(int row, int col){
        if (posOK(row,col)){
            if (labyrinth[nRows+row-1][col] == COMBAT_CHAR){
                labyrinth[nRows+row-1][col] = MONSTER_CHAR;
            }
            else {
                labyrinth[nRows+row-1][col] = EMPTY_CHAR;
            }
        }
    }
    
    /**
     * @brief 
     * @param row
     * @param col
     * @param direction
     * @return 
     */
    private int[] dir2Pos (int row, int col, Directions direction){
        
        int resultado[] = new int[2];
        switch (direction) {
            case DOWN:
                resultado[0] = row - 1;
                resultado[1] = col;
                break;
            case UP:
                resultado[0] = row + 1;
                resultado[1] = col;
                break;
            case LEFT:
                resultado[0] = row;
                resultado[1] = col - 1;
                break;
            case RIGHT:
                resultado[0] = row;
                resultado[1] = col + 1;
                break;
        }
        
        return resultado;
    }
    
    /**
     * @brief Genera posición aleatoria del laberinto aseguranfo que esté vacía 
     * @return la posición
     */
    private int[] randomEmpyPos (){
        
        Dice dice = new Dice();
        int fila, columna;
        
        do{
            fila = dice.randomPos(nRows);
            columna = dice.randomPos(nCols);
        }
        while (labyrinth[fila][columna] != EMPTY_CHAR);
        
        int resultado[] = new int[2];
        resultado[0] = fila;
        resultado[1] = columna;
        
        return resultado;
    }
    
    // P3
    private Monster putPlayer2D (int oldRow, int oldCol, int row, int col, Player player){
        throw new UnsupportedOperationException();
    }
    
    // Métodos públicos
    
    public int getNRows(){
        return nRows;
    }
    
    public int getNCols(){
        return nCols;
    }
    
    // P3
    public void spreadPlayers (ArrayList<Player> players){
        throw new UnsupportedOperationException();
    }
    
    /**
     * @brief Devuelve true si hay un jugador en la casilla de salida
     * @return True di lo hay, false en caso contrario
     */
    public boolean haveAWinner(){
        return !(players[nRows+exitRow-1][exitCol] == null);
    }
    
    // To String
    @Override
    public String toString(){
        
        String cadena = "";
        for (int i = nRows-1; i >= 0; i--){
            for (int j = 0; j < nCols; j++){
                cadena += labyrinth[i][j];
                if (j == nCols -1) cadena += "\n";
            }
        }
        
        return cadena;
    }
    
    /**
     * @brief Añade un monstruo al laberinto 
     * @param row Fila del monstruo
     * @param col Columna del monstruo
     * @param monster Monstruo a añadir
     */
    public void addMonster (int row, int col, Monster monster){
        if (posOK(row,col)){
            if (emptyPos(row,col)){
                labyrinth[nRows+row-1][col] = MONSTER_CHAR;
                monsters[nRows+row-1][col] = monster;
                monster.setPos(row, col);
            }
        }
    }
        
    
    public Monster putPlayer (Directions direction, Player player){
        throw new UnsupportedOperationException();
    }
    
    public void addBlock (Orientation orientation, int startRow, int length) {
        throw new UnsupportedOperationException();
    }
    
    public Directions[] validMoves(int row, int col){
        throw new UnsupportedOperationException();
    }
    
    
}

File: Ruby/P2/Irrgarten/Labyrinth.rb
#encoding:utf-8

=begin

    Autor: Isabel Morro Tabares 

    Funciones del Laberinto
    EL (0,0) ES LA POSICIÓN SUPERIOR IZQUIERDA 
    LA (NROWS-1, NCOLS-1) ES LA POSCIÓN INFERIOR DERECHA
    
=end

require_relative 'Dice'
require_relative 'Monster'
require_relative 'Shield'
require_relative 'Orientation'
require_relative 'Directions'

module Irrgarten
    class Labyrinth

    private

        # Atributos de clase

        @@BLOCK_CHAR = 'X'
        @@EMPTY_CHAR = '-'
        @@MONSTER_CHAR = 'M'
        @@COMBAT_CHAR = 'C'
        @@EXIT_CHAR = 'E'
        @@ROW = 0
        @@COL = 1

        # Métodos privados

        def posOK (row, col)
            row < @nRows && col < @nCols && row >= 0 && col >= 0
        end

        def emptyPos (row, col)
            @labyrinth[row][col] == @@EMPTY_CHAR
        end

        def monsterPos(row, col)
            @labyrinth[row][col] == @@MONSTER_CHAR
        end

        def exitPos (row, col)
            @labyrinth[row][col] == @@EXIT_CHAR
        end

        def combatPos (row, col)
            @labyrinth[row][col] == @@COMBAT_CHAR
        end

        def canStepOn (row, col)
            posOK(row, col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col))
        end

        def updateOldPos (row, col)
            if posOK(row,col)
                if @labyrinth[row][col] == COMBAT_CHAR
                    @labyrinth[row][col] = MONSTER_CHAR
                else 
                    @labyrinth[row][col] = EMPTY_CHAR
                end
            end
        end

        def dir2Pos (row, col, direction)
            
            resultado[] = Array.new

            case direction
                when DOWN
                    resultado[0] = row - 1
                    resultado[1] = col
                when UP
                    resultado[0] = row + 1
                    resultado[1] = col
                when LEFT
                    resultado[0] = row
                    resultado[1] = col - 1
                else RIGHT
                    resultado[0] = row
                    resultado[1] = col + 1 
            end
            
            resultado
        end

        def randomemptyPos

            dice = Dice.new

            begin
                fila = dice.randomPost(@nRows)
                columna = dice.randomPost(@nCols)
            end while @@labyrinth[fila][columna] != @@EMPTY_CHAR

            resultado = Array.new
            resultado[0] = fila
            resultado[1] = columna
        end

        # P3
        def putPlayer2D (oldRow, oldCol, row, col, player)
        end

    public

        # Constructor

        def initialize (nRows, nCols, exitRow, exitCol)

            @nRows = nRows
            @nCols = nCols
            @exitRow = exitRow
            @exitCol = exitCol

            @monsters = Array.new(@nRows)
            for i in (0..@nRows-1)
                @monsters[i] = Array.new(@nCols)
            end
            @players = Array.new(@nRows)
            for i in (0..@nRows-1)
                @players[i] = Array.new(@nCols)
            end
            @labyrinth = Array.new(@nRows)
            for i in (0..@nRows-1)
                @labyrinth[i] = Array.new(@nCols)
            end

            for i in (0..@nRows-1) 
                for j in  (0..@nCols-1)
                    if (i == 0 || i == @nRows-1 || j == 0 || j == @nCols-1)
                        @labyrinth[i][j] = @@BLOCK_CHAR
                    else
                        @labyrinth[i][j] = @@EMPTY_CHAR
                    end 
                end 
            end

            @labyrinth[@exitRow][@exitCol] = @@EXIT_CHAR

        end

        # Métodos públicos

        def nRows
            @nRows
        end 

        def nCols
            @nCols
        end

        # P3
        def spreadPlayers (players)
        end

        def haveAWinner
            !(@players[@exitRow][@exitCol] == nil)
        end

        def to_s
            cadena = ""
            for i in (0..@nRows-1)
                for j in (0..@nCols-1)
                    cadena += @labyrinth[i][j]
                    if j == @nCols-1
                        cadena += "\n"
                    end
                end
            end
            cadena
        end

        def addMonster (row, col, monster)
            if posOK(row,col)
                if emptyPos(row,col)
                    @labyrinth[row][col] = @@MONSTER_CHAR
                    @monsters[row][col] = monster
                    monster.setPos(row, col)
                end
            end
        end

        # P3
        def putPlayer (direction, player)
        end

        # P3
        def addBlock(orientation, startRow, startCol, length)
        end

        # P3
        def validMoves(row, col)
        end
    end
end